sdk/core/azure_core_amqp/src/messaging.rs

```diff
@@ -188,6 +190,121 @@ impl AmqpMessageHeader {
+        let field_count = list.len();
+        if field_count >= 1 {
+            match list.0[0] {
+                AmqpValue::Boolean(durable) => {
+                    builder = builder.with_durable(durable);
+                }
+                _ => {}
+            }
+        }
+        if field_count >= 2 {
+            match list.0[1] {
+                AmqpValue::UByte(priority) => {
+                    builder = builder.with_priority(priority);
+                }
+                _ => {}
+            }
+        }
+        if field_count >= 3 {
+            match list.0[2] {
+                AmqpValue::UInt(time_to_live) => {
+                    builder = builder
+                        .with_time_to_live(std::time::Duration::from_millis(time_to_live.into()));
+                }
+                _ => {}
+            }
+        }
+        if field_count >= 4 {
+            match list.0[3] {
+                AmqpValue::Boolean(first_acquirer) => {
+                    builder = builder.with_first_acquirer(first_acquirer);
+                }
+                _ => {}
+            }
+        }
+        if field_count >= 5 {
+            match list.0[4] {
+                AmqpValue::UInt(delivery_count) => {
+                    builder = builder.with_delivery_count(delivery_count);
+                }
+                _ => {}
+            }
+        }
+        builder.build()
+    }
+}
+
+impl From<AmqpMessageHeader> for AmqpList {
+    fn from(header: AmqpMessageHeader) -> AmqpList {
+        let mut list = Vec::new();
+
+        // Serialize the current value, if it exists. Otherwise serialize a null
+        // value if there are other values to serialize.
+        if let Some(durable) = header.durable {
+            list.push(AmqpValue::Boolean(durable));
+        } else if header.priority.is_some()
+            || header.time_to_live.is_some()
+            || header.first_acquirer.is_some()
+            || header.delivery_count.is_some()
+        {
+            list.push(AmqpValue::Null);
+        }
+        if let Some(priority) = header.priority {
+            list.push(AmqpValue::UByte(priority));
+        } else if header.time_to_live.is_some()
+            || header.first_acquirer.is_some()
+            || header.delivery_count.is_some()
+        {
+            list.push(AmqpValue::Null);
+        }
+        if let Some(time_to_live) = header.time_to_live {
+            list.push(AmqpValue::UInt(time_to_live.as_millis() as u32));
+        } else if header.first_acquirer.is_some() || header.delivery_count.is_some() {
+            list.push(AmqpValue::Null);
+        }
+        if let Some(first_acquirer) = header.first_acquirer {
+            list.push(AmqpValue::Boolean(first_acquirer));
+        } else if header.delivery_count.is_some() {
+            list.push(AmqpValue::Null);
+        }
+        if let Some(delivery_count) = header.delivery_count {
+            list.push(AmqpValue::UInt(delivery_count));
+        }
```

Comment:
There's a lot of duplication here. Might be cleaner to do something like:

```rust
let list = vec![AmqpValue::Null; 5];
if let Some(durable) = header.durable {
    list[0] = durable;
}
if let Some(priority) = header.priority {
    list[1] = priority;
}
// etc.
```
Also, if `durable` is `None`, don't you need to also push `Null` or is that not how the protocol works?

Working e.g.,
```rust
#[allow(dead_code)]
#[derive(Clone, Debug)]
enum AmqpValue {
    Null,
    Boolean(bool),
    U32(u32),
}

struct Options {
    foo: Option<bool>,
    bar: Option<u32>,
}

fn main() {
    let opts = Options {
        foo: Some(true),
        bar: None,
    };

    let mut list = vec![AmqpValue::Null; 2];
    if let Some(foo) = opts.foo {
        list[0] = AmqpValue::Boolean(foo);
    }
    if let Some(bar) = opts.bar {
        list[1] = AmqpValue::U32(bar);
    }

    println!("{list:#?}");
}
```

Reply:
Same - even more so, actually - below.

Reply:
If durable is None, you need to push Null onto the array - . And you need to trim all the trailing null fields.

The array is mutable though. I can resize the array to remove all the trailing nulls.

Reply:
Ah, so `Null` is terminal? Shouldn't there be a return then to short-circuit the method? Or did I misunderstand?

Reply:
No, `Null` is explicitly *not* terminal. But each array slot has a particular semantic - for instance, `first_acquirer` is always at index 3 in the array. If `first_acquirer` is the only field, the array will have 4 entries, 3 of them are `Null`, the 4th will be a `boolean` value. The last entry in the array needs to be the last entry with a value.

Reply:
And I have a much better pattern for this - serialize the whole thing to an array, then scan backwards to the first non Null value and resize the array to remove them.

It makes the conditionals dramatically simpler.